subprojects/model-core/src/main/java/org/gradle/api/internal/provider/FilteringProvider.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.InvalidUserCodeException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.Predicate;
+
+/**
+ * A filtering provider that uses a predicate to filter the value of another provider.
+ **/
+public class FilteringProvider<T> extends AbstractMinimalProvider<T> {
```

Comment:
:thinking: it looks very similar to the `TransformBackedProvider`. I wonder if a base class for these two or other way (have a parameter for `mapValue`/`filterValue`) to reduce duplication is justified.

Reply:
:thinking: https://github.com/gradle/gradle/blob/4d539dedd8294f4cd901bf35e7d91d9c52c2bc4a/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/TransformBackedProvider.java#L30-L32

It's also strange to see this check being added here while we don't have one in the `zip` provider, for example, which also runs user code against the current value of providers.

Reply:
@mlopatkin I thought about it before filing a PR but decided that it was not worth it - the difference is subtle. However, this can be addressed later with more general scope as Rodrigo pointed out - together with zip and flatMap.